toolchain/check/eval.cpp

```diff
@@ -372,28 +375,57 @@ static auto GetConstantValue(EvalContext& eval_context, SemIR::InstId inst_id,
   return eval_context.constant_values().GetInstId(const_id);
 }
 
-// If the given instruction is constant, returns its constant value. When
-// determining the phase of the result, ignore any dependence on `.Self`.
+// Gets a constant value for an `inst_id`, diagnosing when the input is not a
+// constant value.
+static auto RequireConstantValue(EvalContext& eval_context,
+                                 SemIR::InstId inst_id, Phase* phase)
+    -> SemIR::InstId {
+  if (!inst_id.has_value()) {
+    return SemIR::InstId::None;
+  }
+  auto const_id = eval_context.GetConstantValue(inst_id);
+  *phase =
+      LatestPhase(*phase, GetPhase(eval_context.constant_values(), const_id));
+  if (const_id.is_constant()) {
+    return eval_context.constant_values().GetInstId(const_id);
+  }
+
+  if (inst_id != SemIR::ErrorInst::InstId) {
+    CARBON_DIAGNOSTIC(
+        EvalRequiresConstantValue, Error,
+        "found runtime value where a compile-time value was expected");
```

Comment:
This seems kind of borderline for diagnostic phrasing since it's stating the expectation (versus [diagnostic messaging](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide), "Diagnostics should describe the situation the toolchain observed. The language rule violated can be mentioned if it wouldn't otherwise be clear.")

At least I'd suggest something that drops things like "a" and "was", e.g.:

```suggestion
        "expression is runtime; expected constant");
````

Note partly avoiding compile-time, per https://github.com/carbon-language/carbon-lang/pull/2964

But, not really strongly opinionated here -- I think we're going to have to go back and rephrase diagnostics in general